File: packages/python/pii_vault/providers/openai.py
from typing import Iterator, Optional

from ..detector import Detector
from ..tokenizer import Tokenizer
from ..vault import LocalVault


class SafeOpenAI:
    """
    Drop-in replacement for openai.OpenAI.

    Local vault (default)::

        client = SafeOpenAI(api_key="sk-...", vault_key="my-secret")

    Hosted vault (Mawlaia cloud)::

        client = SafeOpenAI(
            api_key="sk-...",
            vault_key="my-hmac-secret",
            mawlaia_api_key="mwl_live_...",
        )

    Or bring your own vault::

        client = SafeOpenAI(api_key="sk-...", vault_key="...", vault=MyVault())
    """

    def __init__(
        self,
        api_key:          str,
        vault_key:        str = "",
        vault_path:       str = ":memory:",
        mawlaia_api_key:  Optional[str] = None,
        mawlaia_vault_url: str = "https://api.mawlaia.com",
        vault=None,
        entities:         Optional[list[str]] = None,
        llm_fallback:     bool = False,
        token_mode:       str = "typed",
        **openai_kwargs,
    ):
        from openai import OpenAI

        self._raw = OpenAI(api_key=api_key, **openai_kwargs)

        if vault is not None:
            _vault = vault
        elif mawlaia_api_key:
            from ..hosted_vault import HostedVault
            _vault = HostedVault(api_key=mawlaia_api_key, vault_url=mawlaia_vault_url)
        else:
            _vault = LocalVault(path=vault_path)

        detector  = Detector(
            entities=entities,
            llm_fallback=llm_fallback,
            openai_api_key=api_key if llm_fallback else None,
        )
        tokenizer = Tokenizer(vault=_vault, key=vault_key or "default", mode=token_mode)

        self.chat   = _ChatNamespace(self._raw, detector, tokenizer)
        self.models = self._raw.models

    def __getattr__(self, name: str):
        return getattr(self._raw, name)


class _ChatNamespace:
    def __init__(self, client, detector: Detector, tokenizer: Tokenizer):
        self.completions = _Completions(client, detector, tokenizer)


class _Completions:
    def __init__(self, client, detector: Detector, tokenizer: Tokenizer):
        self._client    = client
        self._detector  = detector
        self._tokenizer = tokenizer

    def create(self, *, messages: list[dict], subject_id: Optional[str] = None, **kwargs):
        tokenized = self._tokenizer.tokenize_messages(
            messages, self._detector, subject_id=subject_id
        )

        if kwargs.get("stream"):
            return _StreamingResponse(
                self._client.chat.completions.create(messages=tokenized, **kwargs),
                self._tokenizer,
            )

        response = self._client.chat.completions.create(messages=tokenized, **kwargs)

        for choice in response.choices:
            if choice.message and choice.message.content:
                choice.message.content = self._tokenizer.dehydrate(choice.message.content)

        return response


class _StreamingResponse:
    """
    Wraps a streaming OpenAI response.
    Buffers output to avoid yielding partial tokens, dehydrates complete text.
    """

    def __init__(self, stream, tokenizer: Tokenizer):
        self._stream    = stream
        self._tokenizer = tokenizer
        self._buffer    = ""

    def __iter__(self) -> Iterator:
        for chunk in self._stream:
            delta = chunk.choices[0].delta if chunk.choices else None
            if delta and delta.content:
                self._buffer += delta.content
                safe, self._buffer = self._tokenizer.split_stream_safe(self._buffer)
                if safe:
                    chunk.choices[0].delta.content = self._tokenizer.dehydrate(safe)
                    yield chunk
            else:
                yield chunk

        # Flush remaining buffer in a final synthetic yield
        if self._buffer:
            # Build a minimal chunk-like object to carry the flushed content
            import copy
            final = copy.deepcopy(chunk)  # type: ignore[possibly-unbound]
            final.choices[0].delta.content = self._tokenizer.dehydrate(self._buffer)
            self._buffer = ""
            yield final

    def __enter__(self):
        return self

    def __exit__(self, *args):
        if hasattr(self._stream, "__exit__"):
            self._stream.__exit__(*args)

File: packages/typescript/src/providers/openai.ts
import { Detector } from "../detector";
import { Message } from "../models";
import { Tokenizer } from "../tokenizer";
import { Vault } from "../vault";

export interface SafeOpenAIOptions {
  apiKey:      string;
  vaultKey:    string;
  vaultPath?:  string;
  entities?:   string[];
  tokenMode?:  "typed" | "opaque";
  [key:        string]: unknown;
}

export class SafeOpenAI {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private raw:       any;
  private detector:  Detector;
  private tokenizer: Tokenizer;
  public  chat:      ChatNamespace;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public  models:    any;

  constructor(options: SafeOpenAIOptions) {
    const { apiKey, vaultKey, vaultPath = ":memory:", entities, tokenMode = "typed", ...rest } = options;

    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const { default: OpenAI } = require("openai");
    this.raw       = new OpenAI({ apiKey, ...rest });
    const vault    = new Vault(vaultPath);
    this.detector  = new Detector({ entities });
    this.tokenizer = new Tokenizer(vault, vaultKey, tokenMode);
    this.chat      = new ChatNamespace(this.raw, this.detector, this.tokenizer);
    this.models    = this.raw.models;
  }
}

class ChatNamespace {
  completions: Completions;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  constructor(client: any, detector: Detector, tokenizer: Tokenizer) {
    this.completions = new Completions(client, detector, tokenizer);
  }
}

class Completions {
  constructor(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    private client:    any,
    private detector:  Detector,
    private tokenizer: Tokenizer,
  ) {}

  async create({
    messages,
    subjectId,
    ...kwargs
  }: { messages: Message[]; subjectId?: string; [key: string]: unknown }) {
    const tokenized = this.tokenizer.tokenizeMessages(messages, this.detector, subjectId);

    if (kwargs["stream"]) {
      const stream = await this.client.chat.completions.create({ messages: tokenized, ...kwargs });
      return new StreamingResponse(stream, this.tokenizer);
    }

    const response = await this.client.chat.completions.create({ messages: tokenized, ...kwargs });

    for (const choice of response.choices) {
      if (choice.message?.content) {
        choice.message.content = this.tokenizer.dehydrate(choice.message.content as string);
      }
    }

    return response;
  }
}

class StreamingResponse {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  constructor(private stream: any, private tokenizer: Tokenizer) {}

  async *[Symbol.asyncIterator]() {
    let buffer = "";
    let lastChunk: unknown;

    for await (const chunk of this.stream) {
      lastChunk = chunk;
      const content: string | undefined = chunk.choices?.[0]?.delta?.content;

      if (content) {
        buffer += content;
        const [safe, remainder] = this.tokenizer.splitStreamSafe(buffer);
        buffer = remainder;
        if (safe) {
          chunk.choices[0].delta.content = this.tokenizer.dehydrate(safe);
          yield chunk;
        }
      } else {
        yield chunk;
      }
    }

    if (buffer) {
      yield { choices: [{ delta: { content: this.tokenizer.dehydrate(buffer) }, finish_reason: "stop" }] };
    }

    void lastChunk; // used only to satisfy TS
  }
}
